### PR TITLE
OSCD: adding configuration file

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

**Review by calebrob6 — DISMISSED** [review]

**Comment by calebrob6** [comment]
@iejMac looks good, is this done?

**Comment by iejMac** [comment]
Oh no, I haven't even tested this at all yet. Waiting on the random crop sampling PR first

**Comment by calebrob6** [comment]
FYI need to remove `conf/.oscd.yaml.swp`

File: conf/oscd.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+program:
```
Comment (calebrob6) [inline conf/oscd.yaml:1]: can you remove the `program` key

File: conf/oscd.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  min_epochs: 20
+  max_epochs: 500
+  benchmark: True
+
+experiment:
+  task: "oscd"
+  name: "oscd_test5"
+  module:
+    loss: "jaccard"
+    segmentation_model: "fcn"
```
Comment (calebrob6) [inline conf/oscd.yaml:19]: default to unet maybe?

File: conf/task_defaults/oscd.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+experiment:
+  task: "oscd"
+  module:
+    loss: "jaccard"
+    segmentation_model: "unet"
+    encoder_name: "resnet18"
+    encoder_weights: "imagenet"
```
Comment (calebrob6) [inline conf/task_defaults/oscd.yaml:7]: null instead of "imagenet"

File: conf/oscd.yaml (hunk 2 above)
Comment (iejMac) [reply conf/oscd.yaml]: whoops, accident

File: torchgeo/datasets/oscd.py [hunk 4]
```diff
@@ -406,10 +407,49 @@ def __init__(
             K.RandomCrop(patch_size), data_keys=["input", "mask"], same_on_batch=True
         )
 
+    def pad_to(
```
Comment (isaaccorley) [inline torchgeo/datasets/oscd.py:410]: I'm not sure why this can't simply reuse kornia.augmentation.PadTo and apply it to both the image and mask separately.

File: torchgeo/datasets/oscd.py [hunk 5]
```diff
@@ -432,17 +472,19 @@ def setup(self, stage: Optional[str] = None) -> None:
...
+                masks.append(mask.long())
             sample["image"] = torch.stack(images)
             sample["mask"] = torch.stack(masks)
             return sample
 
         train_transforms = Compose([self.preprocess, n_random_crop])
-        test_transforms = Compose([self.preprocess])
+        # for testing and validation we pad all inputs to a fixed size to avoid issues
+        # with the upsampling paths in encoder-decoder architectures
+        test_transforms = Compose([self.preprocess, self.pad_to(1024 + 256)])
```
Comment (isaaccorley) [inline torchgeo/datasets/oscd.py:487]: Your train set returns patches but the test and val set don't. How are you planning to feed this data to a model for eval even though it hasn't seen patches that large during training?

**Review by isaaccorley — CHANGES_REQUESTED** [review]
I think for the most part this is good. My only other thought would be that the we should make the custom collate_fn for the train dataloader a separate fn in the script in case anyone wants to reuse it outside kf the datamodule.

File: torchgeo/datasets/oscd.py (hunk 4 above)
Comment (calebrob6) [reply torchgeo/datasets/oscd.py:410]: @iejMac can you check if this works instead of the custom padto?

File: torchgeo/datasets/oscd.py (hunk 5 above)
Comment (calebrob6) [reply torchgeo/datasets/oscd.py:487]: @isaaccorley for fully convolutional networks this shouldn't be a problem, right? i.e. there is nothing in a U-Net that depends on the size of the image

Comment (isaaccorley) [reply torchgeo/datasets/oscd.py:487]: That's correct. Now that I look at the image sizes I don't think this is a problem since they are generally around 512x512. For some reason I thought the image sizes varied more but it looks like only a few images are above 1000x1000.

I guess my only concern now is we are allowing a user to defined crop size but then hardcoding the test set padding to 1280x1280?


Here are the image sizes for clarity:

```python
# Train set
(785, 799)
(525, 471)
(772, 902)
(1070, 1180)
(360, 395)
(461, 517)
(788, 1015)
(540, 695)
(557, 858)
(582, 522)
(390, 408)
(718, 776)
(563, 339)
(688, 639)

# Test Set 
(469, 433)
(544, 730)
(634, 774)
(716, 824)
(558, 545)
(451, 426)
(385, 241)
(426, 353)
(688, 639)
(476, 458)
```

Comment (calebrob6) [reply torchgeo/datasets/oscd.py:487]: What is your concern above? I picked 1280 because it is > all test set sizes and evenly divisible by 2 enough times to satisfy a U-Net.

Comment (isaaccorley) [reply torchgeo/datasets/oscd.py:487]: No concern then. Just interested to see how we handle that in the trainer.

File: torchgeo/datasets/oscd.py (hunk 4 above)
Comment (iejMac) [reply torchgeo/datasets/oscd.py]: Works great

**Review by calebrob6 — DISMISSED** [review]
Looks good to me, @isaaccorley do you mind checking this too?

File: torchgeo/datasets/oscd.py [hunk 6]
```diff
@@ -432,17 +432,25 @@ def setup(self, stage: Optional[str] = None) -> None:
...
                 mask = mask.squeeze()[0]
                 images.append(image.squeeze())
-                masks.append(mask)
+                masks.append(mask.long())
             sample["image"] = torch.stack(images)
             sample["mask"] = torch.stack(masks)
             return sample
 
+        def pad_to(sample: Dict[str, Any]) -> Dict[str, Any]:
+            padto = K.PadTo((1280, 1280))
```
Comment (isaaccorley) [inline torchgeo/datasets/oscd.py:445]: This is instantiating a K.PadTo object on each call. We should make this a class attribute.

**Review by isaaccorley — CHANGES_REQUESTED** [review]

**Review by isaaccorley — APPROVED** [review]
Lgtm

**Review by adamjstewart — APPROVED** [review]
